[FLINK-31806] Fix public API rule for connectors

echauchot opened the pull request.

echauchot commented:
Tested it on Cassandra connector and it reports these violations:

1. If I put back IOUtils use in the connector it raises 

`Method <org.apache.flink.connector.cassandra.source.enumerator.CassandraEnumeratorStateSerializer.deserialize(int, [B)> calls method <org.apache.flink.util.IOUtils.readFully(java.io.InputStream, [B, int, int)> in (CassandraEnumeratorStateSerializer.java:90)
`
=> This is what we wanted

2. On current code base, it raises:
`java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'Connector production code must not depend on non-public API outside of connector packages' was violated (16 times):
Constructor <org.apache.flink.connector.cassandra.source.CassandraSource.<init>(org.apache.flink.streaming.connectors.cassandra.ClusterBuilder, long, java.lang.Class, java.lang.String, org.apache.flink.streaming.connectors.cassandra.MapperOptions)> calls method <org.apache.flink.api.java.ClosureCleaner.clean(java.lang.Object, org.apache.flink.api.common.ExecutionConfig$ClosureCleanerLevel, boolean)> in (CassandraSource.java:138)
Constructor <org.apache.flink.connector.cassandra.source.CassandraSource.<init>(org.apache.flink.streaming.connectors.cassandra.ClusterBuilder, long, java.lang.Class, java.lang.String, org.apache.flink.streaming.connectors.cassandra.MapperOptions)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object, java.lang.String)> in (CassandraSource.java:124)
Constructor <org.apache.flink.connector.cassandra.source.CassandraSource.<init>(org.apache.flink.streaming.connectors.cassandra.ClusterBuilder, long, java.lang.Class, java.lang.String, org.apache.flink.streaming.connectors.cassandra.MapperOptions)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object, java.lang.String)> in (CassandraSource.java:125)
Constructor <org.apache.flink.connector.cassandra.source.CassandraSource.<init>(org.apache.flink.streaming.connectors.cassandra.ClusterBuilder, long, java.lang.Class, java.lang.String, org.apache.flink.streaming.connectors.cassandra.MapperOptions)> calls method <org.apache.flink.util.Preconditions.checkNotNull(java.lang.Object, java.lang.String)> in (CassandraSource.java:126)
Constructor <org.apache.flink.connector.cassandra.source.CassandraSource.<init>(org.apache.flink.streaming.connectors.cassandra.ClusterBuilder, long, java.lang.Class, java.lang.String, org.apache.flink.streaming.connectors.cassandra.MapperOptions)> calls method <org.apache.flink.util.Preconditions.checkState(boolean, java.lang.String, [Ljava.lang.Object;)> in (CassandraSource.java:127)
Constructor <org.apache.flink.connector.cassandra.source.CassandraSource.<init>(org.apache.flink.streaming.connectors.cassandra.ClusterBuilder, long, java.lang.Class, java.lang.String, org.apache.flink.streaming.connectors.cassandra.MapperOptions)> gets field <org.apache.flink.api.common.ExecutionConfig$ClosureCleanerLevel.RECURSIVE> in (CassandraSource.java:138)
Field <org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializer.SERIALIZER_CACHE> has generic type <java.lang.ThreadLocal<org.apache.flink.core.memory.DataOutputSerializer>> with type argument depending on <org.apache.flink.core.memory.DataOutputSerializer> in (CassandraSplitSerializer.java:0)
Method <org.apache.flink.connector.cassandra.source.CassandraSource.checkQueryValidity(java.lang.String)> calls method <org.apache.flink.util.Preconditions.checkState(boolean, java.lang.Object)> in (CassandraSource.java:145)
Method <org.apache.flink.connector.cassandra.source.CassandraSource.checkQueryValidity(java.lang.String)> calls method <org.apache.flink.util.Preconditions.checkState(boolean, java.lang.Object)> in (CassandraSource.java:149)
Method <org.apache.flink.connector.cassandra.source.CassandraSource.checkQueryValidity(java.lang.String)> is annotated with <org.apache.flink.annotation.VisibleForTesting> in (CassandraSource.java:0)
Method <org.apache.flink.connector.cassandra.source.reader.CassandraSplitReader.generateRangeQuery(java.lang.String, java.lang.String)> is annotated with <org.apache.flink.annotation.VisibleForTesting> in (CassandraSplitReader.java:0)
Method <org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializer.deserialize(int, [B)> calls constructor <org.apache.flink.core.memory.DataInputDeserializer.<init>([B)> in (CassandraSplitSerializer.java:57)
Method <org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializer.serialize(org.apache.flink.connector.cassandra.source.split.CassandraSplit)> calls method <org.apache.flink.core.memory.DataOutputSerializer.clear()> in (CassandraSplitSerializer.java:51)
Method <org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializer.serialize(org.apache.flink.connector.cassandra.source.split.CassandraSplit)> calls method <org.apache.flink.core.memory.DataOutputSerializer.getCopyOfBuffer()> in (CassandraSplitSerializer.java:50)
Method <org.apache.flink.connector.cassandra.source.split.SplitsGenerator.estimateTableSize()> is annotated with <org.apache.flink.annotation.VisibleForTesting> in (SplitsGenerator.java:0)
Static Initializer <org.apache.flink.connector.cassandra.source.split.CassandraSplitSerializer.<clinit>()> calls constructor <org.apache.flink.core.memory.DataOutputSerializer.<init>(int)> in (CassandraSplitSerializer.java:34)
`
=> These are calls to ClosureCleaner, checkState and checkNotNull, VisibleForTesting which should be added as violation exceptions
=> The only discutable part is the dependency on `org.apache.flink.core.memory.DataInputDeserializer` and `org.apache.flink.core.memory.DataOutputSerializer` which are internal APIs but we could definitely use [ByteArray | Object]Streams JDK counterparts instead in the Cassandra code.

zentol reviewed (commented):
So why didn't the previous version work? Is `resideOutsideOfPackages` just a filter that works on a previously provided set of classes, and since we didn't specify anything like `resideInAPackage` it filtered on an empty set?

zentol commented:
> => These are calls to ClosureCleaner, checkState and checkNotNull, VisibleForTesting which should be added as violation exceptions
=> The only discutable part is the dependency on org.apache.flink.core.memory.DataInputDeserializer and org.apache.flink.core.memory.DataOutputSerializer which are internal APIs but we could definitely use [ByteArray | Object]Streams JDK counterparts instead in the Cassandra code.

I'd start a separate discussion on these. we likely want to make Preconditions/VisibleForTesting public.
Not sure about `DataInput*`, maybe we can look at how it would look like in Cassandra and then make a call.
ClosureCleaner I'm not sure about, but we could consider handling this internally in Flink instead of having each connector use it (e.g., call it in StreamExEnv#addSource)

echauchot commented:
> So why didn't the previous version work? Is `resideOutsideOfPackages` just a filter that works on a previously provided set of classes, and since we didn't specify anything like `resideInAPackage` it filtered on an empty set?

No, it did not work because the rule was conceptually wrong: see the review comment I have put on line 50

echauchot commented:
> I'd start a separate discussion on these. we likely want to make Preconditions/VisibleForTesting public. Not sure about `DataInput*`, maybe we can look at how it would look like in Cassandra and then make a call. ClosureCleaner I'm also not sure about, but we could consider handling this internally in Flink instead of having each connector use it (e.g., call it in StreamExEnv#addSource)

Yes I was about to open a dedicated Cassandra ticket for `DataInput*` / `DataOutput*` removal. I could open a PR to replace them with JDK counterparts and discuss there if it is better for Cassandra connector to use JDK libs instead of non-public flink lib.

echauchot commented:
> > Not sure about `DataInput*`, maybe we can look at how it would look like in Cassandra and then make a call.
> 
> Yes I was about to open a dedicated Cassandra ticket for `DataInput*` / `DataOutput*` removal. I could open a PR to replace them with JDK counterparts and discuss there if it is better for Cassandra connector to use JDK libs instead of non-public flink lib.

https://issues.apache.org/jira/browse/FLINK-32222

echauchot commented:
> > No, it did not work because the rule was conceptually wrong: see the review comment I have put on line 50
> 
> I don't see any comment; did you submit the review?
My bad, forgot to submit the review. I thought it was automatic, this PR being my own code

echauchot commented:
I don't get how the centralized flink-architecture-tests-production work. The [azure CI](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=49514&view=logs&j=fc5181b0-e452-5c8f-68de-1097947f6483&t=995c650b-6573-581c-9ce6-7ad4cc038461) complains about `freeze.store.default.allowStoreUpdate` being disabled in `flink-architecture-tests-production` whereas it is enabled in `flink-architecture-tests-production/src/test/resources/archunit.properties`. Even if I enable `freeze.refreeze` in this module as stated in the README, I still get no error locally and no violation addition.

echauchot commented:
> I don't get how the centralized flink-architecture-tests-production work. The [azure CI](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=49514&view=logs&j=fc5181b0-e452-5c8f-68de-1097947f6483&t=995c650b-6573-581c-9ce6-7ad4cc038461) complains about `freeze.store.default.allowStoreUpdate` being disabled in `flink-architecture-tests-production` whereas it is enabled in `flink-architecture-tests-production/src/test/resources/archunit.properties`. Even if I enable `freeze.refreeze` in this module as stated in the README, I still get no error locally and no violation addition.

I have to recreate the violation store, just updating it did not work

echauchot commented:
> > > Not sure about `DataInput*`, maybe we can look at how it would look like in Cassandra and then make a call.
> > 
> > 
> > Yes I was about to open a dedicated Cassandra ticket for `DataInput*` / `DataOutput*` removal. I could open a PR to replace them with JDK counterparts and discuss there if it is better for Cassandra connector to use JDK libs instead of non-public flink lib.
> 
> https://issues.apache.org/jira/browse/FLINK-32222

Addressed in Cassandra here: https://github.com/apache/flink-connector-cassandra/pull/16

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

zentol commented on flink-architecture-tests/flink-architecture-tests-production/src/main/java/org/apache/flink/architecture/rules/ConnectorRules.java:
```diff
@@ -41,16 +39,14 @@ public class ConnectorRules {
-                .and(
-                        not(areDirectlyAnnotatedWithAtLeastOneOf(
-                                        Public.class, PublicEvolving.class))
-                                .and(not(modifier(PUBLIC))))
```
Ah not I get it, that makes complete sense.

zentol commented on flink-architecture-tests/flink-architecture-tests-production/archunit-violations/6247d4b0-2dd8-4241-80da-a0d981a8c323:
```diff
@@ -0,0 +1,1486 @@
+Class <org.apache.flink.connector.datagen.table.DataGenVisitorBase$TimeGenerator> implements interface <org.apache.flink.streaming.api.functions.source.datagen.DataGenerator> in (DataGenVisitorBase.java:0)
```
good god...this is gonna be fine to go through...

zentol commented:
Rebasing the PR should resolve the e2e CI issue.

zentol commented on flink-architecture-tests/flink-architecture-tests-production/src/main/java/org/apache/flink/architecture/rules/ConnectorRules.java:
```diff
@@ -41,16 +39,14 @@ public class ConnectorRules {
+            areFlinkClassesThatResideOutsideOfConnectorPackagesAndAreNotPublic() {
+        return JavaClass.Predicates.resideInAPackage("org.apache.flink..")
+                .and(JavaClass.Predicates.resideOutsideOfPackages(CONNECTOR_PACKAGES))
+                .and(not(areDirectlyAnnotatedWithAtLeastOneOf(Public.class, PublicEvolving.class)))
```
We should _somehow_ replicate the logic of also checking the outer class for annotations; right now this creates a lot of false-positives.

Similar to what we do for method checks:
```
                            .areDeclaredInClassesThat(
                                    areJavaClasses()
                                            .and(
                                                    areDirectlyAnnotatedWithAtLeastOneOf(
                                                            PublicEvolving.class)))
```

Not sure yet how to express this properly in archunit.

echauchot replied:
makes sense, coding that

echauchot replied:
thanks, but I got a simpler solution

zentol replied:
yes belongTo should work, at least according to the javadocs.

echauchot commented:
@zentol:
- I rebased on master
- I changed the rule by reverting the assumption from "class must not depend on non-public" to "class must depend only on public". That makes something that I believe is more maintainable and readable. 
- I updated the violations: the rule seems to work and not entail false-positves on non-public classes enclosed in public classes.
PTAL

zentol commented on flink-architecture-tests/flink-architecture-tests-production/archunit-violations/e5126cae-f3fe-48aa-b6fb-60ae6cc3fcd5:
```diff
@@ -7,22 +7,22 @@ Method <org.apache.flink.runtime.blob.BlobInputStream.read()> calls method <org.
 Method <org.apache.flink.runtime.highavailability.nonha.embedded.EmbeddedHaServicesWithLeadershipControl.revokeResourceManagerLeadership()> calls method <org.apache.flink.runtime.highavailability.nonha.embedded.EmbeddedLeaderService.revokeLeadership()> in (EmbeddedHaServicesWithLeadershipControl.java:102)
 Method <org.apache.flink.runtime.io.network.partition.ResourceManagerPartitionTrackerImpl.listDataSets()> calls method <org.apache.flink.runtime.io.network.partition.DataSetMetaInfo.withNumRegisteredPartitions(int, int)> in (ResourceManagerPartitionTrackerImpl.java:286)
-Method <org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator$DeferrableCoordinator.closeAsync(long)> calls method <org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator$QuiesceableContext.quiesce()> in (RecreateOnResetOperatorCoordinator.java:343)
+Method <org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator$DeferrableCoordinator.closeAsync(long)> calls method <org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator$QuiesceableContext.quiesce()> in (RecreateOnResetOperatorCoordinator.java:349)
```
Why did the line numbers change :thinking: I've only seen that happen on Java 17.

echauchot replied:
Also I noticed something worth sharing here:
- using `mvn clean verify` does not update the violation store but running the actual production architecture test does.
- if a rule description changes then the violations with the old rule name are not purged from the store: `sored.rules` index file still contains the old entry and the related violations file is not purged either. I had to remove the violations file manually and remove the entry in the index file when I changed the rule description.

echauchot replied:
@zentol CI passes. Are we good to go ?

zentol replied:
> I guess because I rebased on master and prod code has changed ?

Shouldn't it then fail on master if line numbers weren't updated? I had this issue with Java 17.
But the new numbers are definitely correct.

> using mvn clean verify does not update the violation store but running the actual production architecture test does.

TBH I doubt that. You probably just forgot to set one of the magical archunit properties.

`mvn verify -Darchunit.freeze.refreeze=true` worked fine for me, and I got the same line number changes locally.
Could you move these into a separate hotfix commit? They aren't related to the actual rule fix.

zentol replied:
These commits should be squashed before the merge as otherwise one commit alone would fail CI.
```
[FLINK-31806] Reverse rule to positive asumption
[FLINK-31806] Update violations 
```

echauchot replied:
> > I guess because I rebased on master and prod code has changed ?
> 
> Shouldn't it then fail on master if line numbers weren't updated? I had this issue with Java 17. But the new numbers are definitely correct.

Yes you're right, regular CI on master should fail in that case, I haven't checked but if it were the case we would have seen. TBH I don't know what happens with the line numbers, but I used java 11 to update the violations.

> 
> > using mvn clean verify does not update the violation store but running the actual production architecture test does.
> 
> TBH I doubt that. You probably just forgot to set one of the magical archunit properties.
> 

I did set `archunit.freeze.refreeze=true` with mvn verify only I set it from the archunit.properties file not as System property. 

> `mvn verify -Darchunit.freeze.refreeze=true` worked fine for me, and I got the same line number changes locally.

That is comforting.

 Could you move these into a separate hotfix commit? They aren't related to the actual rule fix.

You mean the line numbers update?

echauchot replied:
> These commits should be squashed before the merge as otherwise one commit alone would fail CI.

done

> Could you move these into a separate hotfix commit? They aren't related to the actual rule fix.

done

PTAL

zentol reviewed (approved):
Since each of these modifies the rules I'd assume each of these individually would also have to update the store.
So it's either that or squash them all together.
(again, key rule is that each merged commit should individually pass on CI; if they'd already pass go ahead)

```
[FLINK-31806] Fix public API rule for connectors
[FLINK-31806] Add a rule to test non-public enclosing classes
[FLINK-31806] Reverse rule to positive asumption and update violation… 
```

echauchot commented:
> Since each of these modifies the rules I'd assume each of these individually would also have to update the store. So it's either that or squash them all together. (again, key rule is that each merged commit should individually pass on CI; if they'd already pass go ahead)
> 
> ```
> [FLINK-31806] Fix public API rule for connectors
> [FLINK-31806] Add a rule to test non-public enclosing classes
> [FLINK-31806] Reverse rule to positive asumption and update violation… 
> ```

Makes total sense. As the first rule was missing the enclosed class case, the second rule was missing a sub-case, there is no point in keeping these commits, I'll squash them.
